internal/gui/gui.go

```diff
@@ -214,212 +193,207 @@ func (ui *GUI) loadTemplates() error {
 		return nil
 	}
 
-	err := filepath.Walk(ui.cfg.GUIDir, findTemplate)
+	err := filepath.Walk(cfg.GUIDir, findTemplate)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
 	httpTemplates := template.New("template").Funcs(template.FuncMap{
 		"upper":          strings.ToUpper,
 		"floatToPercent": floatToPercent,
 	})
 
-	// Since template.Must panics with non-nil error, it is much more
-	// informative to pass the error to the caller to log it and exit
-	// gracefully.
-	httpTemplates, err = httpTemplates.ParseFiles(templates...)
+	return httpTemplates.ParseFiles(templates...)
+}
+
+// NewGUI creates an instance of the user interface.
+func NewGUI(cfg *Config) (*GUI, error) {
+	templates, err := loadTemplates(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	cache, err := initCache(cfg)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	ui.templates = template.Must(httpTemplates, nil)
-	return nil
+	ui := GUI{
+		cfg:             cfg,
+		limiter:         pool.NewRateLimiter(),
+		templates:       templates,
+		cookieStore:     sessions.NewCookieStore(cfg.CSRFSecret),
+		router:          mux.NewRouter(),
+		cache:           cache,
+		websocketServer: NewWebsocketServer(),
+	}
+	ui.route()
+	return &ui, nil
 }
 
-// Run starts the user interface.
-func (ui *GUI) Run(ctx context.Context) {
+// runWebServer starts the web server according per the configuration options
+// associated with the GUI instance.
+//
+// It must be run as a routine.
+func (ui *GUI) runWebServer(ctx context.Context) {
+	// Create base HTTP/S server configuration and start a goroutine to ensure
+	// it is properly shutdown when the context is closed.
+	server := http.Server{
+		// Use the provided context as the parent context for all requests to
+		// ensure handlers are able to react to both client disconnects as well
+		// as shutdown via the provided context.
+		BaseContext: func(l net.Listener) context.Context {
+			return ctx
+		},
+
+		WriteTimeout: time.Second * 30,
+		ReadTimeout:  time.Second * 30,
+		IdleTimeout:  time.Second * 30,
+		Addr:         ui.cfg.GUIListen,
+		Handler:      ui.router,
+	}
 	go func() {
-		switch {
-		case ui.cfg.UseLEHTTPS:
-			certMgr := &autocert.Manager{
-				Prompt:     autocert.AcceptTOS,
-				Cache:      autocert.DirCache("certs"),
-				HostPolicy: autocert.HostWhitelist(ui.cfg.Domain),
-			}
+		<-ctx.Done()
+		server.Shutdown(ctx)
```

Comment:
Afaict from the doco, calling Shutdown with an already canceled context is about as good as not calling it at all. It needs to be given a new context, probably a `context.WithTimeout`, so it has time to do its work. A few seconds is probably adequate for any in-flight requests to complete.

Reply:
I'm fairly certain it's correct and I've also tested that it does work properly.

The call to `Shutdown` is really only necessary to close the listener(s) so the server returns.  In reality, we wouldn't really need to pass the context at all, but it is a required argument.

From the docs, emphasis mine:

> Shutdown gracefully shuts down the server **without interrupting any active connections**. Shutdown works by first **closing all open listeners**, then closing all **idle** connections, and then **waiting indefinitely for connections to return to idle** and then shut down.

> If the provided context expires before the shutdown is complete, Shutdown returns the context's error, otherwise it returns any  error returned from closing the Server's underlying Listener(s).

So, it waits for in-flight requests to complete regardless of context cancellation.  That is why there is also a `BaseContext` which the handlers themselves can use to exit early.

As an aside, I know this to be true from dcrd's RPC server some time ago.  Before many other change and prior to the addition of the `BaseContext`, if you were to kick off a rescan (or any other request that took a long time to complete) and then shutdown dcrd, the http server would hang until the rescan completed because the rescan handler wasn't checking the context and exiting early.

Reply:
I've been [playing around](https://go.dev/play/p/-F2Mrj8ft2Z) with this and you're totally right. The server shuts down cleanly regardless of the status of the passed context. This PR stays as it is, I will update decred/vspd#434.

Reply:
You still might want to log any errors returned by Shutdown?

Reply:
> You still might want to log any errors returned by Shutdown?

I'm not super opposed to it if you feel strongly about, but I generally tend to avoid logging errors the admin can't do anything about anyway.  In my experience, doing that ends up in support tickets.

Reply:
> I've been [playing around](https://go.dev/play/p/-F2Mrj8ft2Z) with this and you're totally right. The server shuts down cleanly regardless of the status of the passed context. This PR stays as it is, I will update [decred/vspd#434](https://github.com/decred/vspd/pull/434).

Here is slightly modified version that shows why the base context is needed too:

https://go.dev/play/p/FbWkwO9DjrR